File: src/core/bitboard/Masks.ts
// Reference: https://github.com/cglouch/snakefish/blob/master/src/tables.py

const ranks: bigint[] = [];
for (let rank = 0; rank < 8; rank += 1)
	ranks.push(0x00000000000000ffn << BigInt(8 * rank));

const files: bigint[] = [];
for (let file = 0; file < 8; file += 1)
	files.push(0x0101010101010101n << BigInt(file));

const queenMasks: bigint[] = [];
const rookMasks: bigint[] = [];
const bishopMasks: bigint[] = [];
const knightMasks: bigint[] = [];

for (let i = 0; i < 64; i += 1) {
	const iBig = BigInt(i);
	const pieceMask = 1n << BigInt(i);

	// STRAIGHT
	const rankMask = ranks[Math.floor(i / 8)];
	const fileMask = files[i % 8];

	const straightAttacks = (rankMask | fileMask) ^ pieceMask;

	// DIAG
	const diag = 8n * (iBig & 7n) - (iBig & 56n);
	const northDiag = -diag & (diag >> 31n);
	const southDiag = diag & (-diag >> 31n);
	const diagMask = (0x8040201008040201n >> southDiag) << northDiag;

	const antiDiag = 56n - 8n * (iBig & 7n) - (iBig & 56n);
	const northAntiDiag = -antiDiag & (antiDiag >> 31n);
	const southAntiDiag = antiDiag & (-antiDiag >> 31n);
	const antiDiagMask =
		(0x0102040810204080n >> southAntiDiag) << northAntiDiag;

	const diagAttacks = (diagMask | antiDiagMask) ^ pieceMask;

	rookMasks.push(straightAttacks);
	bishopMasks.push(diagAttacks);
	queenMasks.push(diagAttacks | straightAttacks);

	// KNIGHT
	const m1 = ~(files[0] | files[1]);
	const m2 = ~files[0];
	const m3 = ~files[7];
	const m4 = ~(files[7] | files[6]);

	const s1 = (pieceMask & m1) << 6n;
	const s2 = (pieceMask & m2) << 15n;
	const s3 = (pieceMask & m3) << 17n;
	const s4 = (pieceMask & m4) << 10n;
	const s5 = (pieceMask & m4) >> 6n;
	const s6 = (pieceMask & m3) >> 15n;
	const s7 = (pieceMask & m2) >> 17n;
	const s8 = (pieceMask & m1) >> 10n;

	knightMasks.push(s1 | s2 | s3 | s4 | s5 | s6 | s7 | s8);
}

export default {
	QUEEN: queenMasks,
	ROOK: rookMasks,
	BISHOP: bishopMasks,
	KNIGHT: knightMasks,
	FILES: files,
	RANKS: ranks
};

File: src/rust/masks.rs
pub struct Masks {
    pub ranks: [u64; 8],
    pub files: [u64; 8],
    pub cell: [u64; 64],
    pub queen: [u64; 64],
    pub rook: [u64; 64],
    pub bishop: [u64; 64],
    pub knight: [u64; 64],
}

// Reference: https://github.com/cglouch/snakefish/blob/master/src/tables.py
pub const fn generate_masks() -> Masks {
    let mut ranks: [u64; 8] = [0; 8];
    let mut files: [u64; 8] = [0; 8];
    let mut idx: usize = 0;

    // For loops cannot be used in const fns -> use while.
    while idx < 8 {
        ranks[idx] = 0x00000000000000FF << (8 * idx);
        files[idx] = 0x0101010101010101 << idx;
        idx += 1
    }

    let mut queen: [u64; 64] = [0; 64];
    let mut rook: [u64; 64] = [0; 64];
    let mut bishop: [u64; 64] = [0; 64];
    let mut knight: [u64; 64] = [0; 64];
    let mut cell: [u64; 64] = [0; 64];

    idx = 0;
    while idx < 64 {
        let piece_mask = 1 << idx as u64;

        let rank_mask = ranks[idx / 8]; // We don't need floor here, int division automatically rounds down
        let file_mask = files[idx % 8];
        let straight_attacks = (rank_mask | file_mask) ^ piece_mask;

        let diag = 8 * (idx as i64 & 7) - (idx as i64 & 56);
        let north_diag = -diag & (diag >> 31);
        let south_diag = diag & (-diag >> 31);
        let diag_mask = (0x8040201008040201 as u64 >> south_diag) << north_diag;

        let anti_diag = 56 - 8 * (idx as i64 & 7) - (idx as i64 & 56);
        let north_anti_diag = -anti_diag & (anti_diag >> 31);
        let south_anti_diag = anti_diag & (-anti_diag >> 31);
        let anti_diag_mask = (0x0102040810204080 as u64 >> south_anti_diag) << north_anti_diag;

        let diag_attacks = (diag_mask | anti_diag_mask) ^ piece_mask;

        rook[idx] = straight_attacks;
        bishop[idx] = diag_attacks;
        queen[idx] = straight_attacks | diag_attacks;
        cell[idx] = piece_mask;

        // KNIGHT
        let m1 = !(files[0] | files[1]);
        let m2 = !files[0];
        let m3 = !files[7];
        let m4 = !(files[7] | files[6]);
        let s1 = (piece_mask & m1) << 6;
        let s2 = (piece_mask & m2) << 15;
        let s3 = (piece_mask & m3) << 17;
        let s4 = (piece_mask & m4) << 10;
        let s5 = (piece_mask & m4) >> 6;
        let s6 = (piece_mask & m3) >> 15;
        let s7 = (piece_mask & m2) >> 17;
        let s8 = (piece_mask & m1) >> 10;

        knight[idx] = s1 | s2 | s3 | s4 | s5 | s6 | s7 | s8;

        idx += 1
    }

    return Masks {
        ranks,
        files,
        cell,
        queen,
        rook,
        bishop,
        knight,
    };
}
